File: 1 Easy/1 - Two Sum/TwoSum.java
/*
 * Given an array of integers, return indices of the two numbers such that they
 * add up to a specific target. You may assume that each input would have
 * exactly one solution.
 */

import java.util.HashMap;

class TwoSum {

    public static void bruteForceSolution(int[] array, int target, int[] ans) {
        for (int i = 0; i < array.length; i++) {
            for (int j = i + 1; j < array.length - 1; i++) {
                if (target == array[i] + array[j]) {
                    ans[0] = i;
                    ans[1] = j;
                    return;
                }
            }
        }
    }

    public static void hashTableTwoPasses(int[] array, int target, int[] ans) {
        HashMap<Integer, Integer> map = new HashMap<>();

        for (int i = 0; i < array.length; i++) {
            map.put(array[i],i);
        }
        for (int i = 0; i < array.length; i++) {
            int complement = target - array[i];
            if (map.containsKey(complement) && map.get(complement) != i) {
                ans[0] = i;
                ans[1] = map.get(complement);
                return;
            }
        }
    }

    public static void hashTableOnePass(int[] array, int target, int[] ans) {
        HashMap<Integer, Integer> map = new HashMap<>();

        for (int i = 0; i < array.length; i++) {
            int complement = target - array[i];

            if (map.containsKey(complement) && map.get(complement) != i) {
                ans[0] = i;
                ans[1] = map.get(complement);
                return;
            }
            map.put(array[i], i);
        }
    }

    public static void main(String[] args) {
        int nums[] = {2, 7, 11, 15};
        int target = 9;
        int ans0[] = {-1, -1};
        int ans1[] = {-1, -1};
        int ans2[] = {-1, -1};

        bruteForceSolution(nums, target, ans0);
        System.out.println("[" + ans0[0] + "," + ans0[1] + "]");

        hashTableTwoPasses(nums, target, ans1);
        System.out.println("[" + ans1[0] + "," + ans1[1] + "]");

        hashTableOnePass(nums, target, ans2);
        System.out.println("[" + ans2[0] + "," + ans2[1] + "]");
    }
}

File: 1 Easy/1 - Two Sum/TwoSum.py
# -*- coding: utf-8 -*-

# Given an array of integers, return indices of the two numbers such that they
# add up to a specific target.
#
# You may assume that each input would have exactly one solution, and you may
# not use the same element twice.
#
# Example
#
# Given nums = [2, 7, 11, 15], target = 9,
# Because nums[0] + nums [1] = 2 + 7 = 9,
# return [0, 1]

class Solution:
    def twoSum(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: List[int]
        """
        res = [-1, -1]
        index = {}
        for i in range(len(nums)):
            complement = target - nums[i]
            if complement in index:
                return [index[complement], i]
            else:
                index[nums[i]] = i
        return res


def main():
    sol = Solution()
    nums = [2, 7, 11, 15]
    target = 9
    print(sol.twoSum(nums, target))

if __name__ == '__main__':
    main()
